### DISCUSSION TITLE
method to delete / update many from database

### DISCUSSION THREAD

**Discussion by ufelgen** [neuefische/questions#450] [Web] [answered]
Hey everyone,

I moved my data from local storage to a database and updated all individual entries that can access an object by ID. 
Now however, I am trying to delete and update many entries at once (deleting a category and updating a category name). 

### 1. PUT 🤯

I think I need the PUT method in the "api/words" route, where I also have the DELETE. However, I don't know what to send as req.body and how to write the updateMany method successfully... 

**Does anyone have an idea how to do this?**

### 2. DELETE - edit: solved 🥳 see below

I almost got this to work using the "api/words" path like this: 

```js
if (req.method === "DELETE") {
    const categoryToDelete = req.body;
    const result = await Word.deleteMany({ category: categoryToDelete });
    if (result) {
      res.status(200).json({ message: "category deleted" });
    } else {
      res.status(404).json({ message: "not found" });
    }
  }
```

the function itself I tried writing like this: 

```js
async function handleDeleteCategory(event, category) {
    event.preventDefault();
    event.stopPropagation();
    const confirmation = confirm(
      "do you wish to delete all words saved in this category?"
    );
    if (confirmation) {
      await fetch("/api/words", {
        method: "DELETE",
        body: JSON.stringify(category),
      });
      async function performFetch() {
        const allWordsFromDatabase = await fetchData();
        onHandleAllWords(allWordsFromDatabase);
      }
      performFetch();
    }
  }
```

However, when I try to pass the category name as req.body, it is wrapped in double quotes:
<img width="356" alt="Bildschirmfoto 2023-01-12 um 18 10 48" src="https://user-images.githubusercontent.com/115288914/212134562-bab10732-2c2b-45cb-ae9a-c600b1a211a9.png">

when I replace `categoryToDelete` with an existing category name (e.g. `const result = await Word.deleteMany({ category: "more" });` then it works. 

**How do I get rid of the double quotes??**

>  **Edit:**
> 
>  **Answer:**
>  **in the function that sets the req.body, I had to replace `body: JSON.stringify(category)` by `body: category`**

Thanks for your help!!

**Comment by ufelgen** [comment, marked as answer]
Cheers, I solved the PUT, too 🥳

the function is 
```js
 async function handleEditedCategory(editId, updatedCategory) {
    await fetch("/api/words", {
      method: "PUT",
      body: [editId, updatedCategory],
    });
    async function performFetch() {
      const allWordsFromDatabase = await fetchData();
      onHandleAllWords(allWordsFromDatabase);
    }
    performFetch();
  }
```

and in the "api/words" I wrote
```js
if (req.method === "PUT") {
    const array = req.body.split(",");
    const [oldCategory, newCategory] = array;
    const updatedCategories = await Word.updateMany(
      { category: oldCategory },
      { category: newCategory }
    );
    if (updatedCategories) {
      res.status(200).json({ message: "category name updated" });
    } else {
      res.status(404).json({ message: "not found" });
    }
  }
```

not sure if the array in the body and then splitting that later is the elegant way. 

**so I'm still thankful for suggestions to improve it** 🙃🙏

**Reply by ThomasKirst** [reply]
Hi Ulli,

in a "perfect world scenario" I would store the category in its own collection and associate the category with the words. Then I would have to only update one document (oldCategoryName –> newCategoryName) and all associated words would automatically be assigned the new category name (by its relation from category <-> words (one-to-many relationship)).

In your example, I would prefer an object with named properties instead of an array:

```js
async function handleEditedCategory(categoryToUpdate, newCategoryName) {
    await fetch("/api/words", {
      method: "PUT",
      body: { categoryToUpdate, newCategoryName },
    });
    // …
```

```js
  // api/words
  if (req.method === "PUT") {
    const { categoryToUpdate, newCategoryName } = req.body;
    const updatedWords = await Word.updateMany(
      { category: categoryToUpdate },
      { category: newCategoryName }
    );
    // …
```

Apart from this, great that you found a solution which works for you :)
– Thomas

**Reply by ufelgen** [reply]
Thank you! :)
